Introduce Action and SceneState

yck011522 opened the pull request:
This pull request is to introduce the Action and State Classes for modeling a robotic fabrication process. The concept of Actions and States is based on this paper https://dl.acm.org/doi/abs/10.1145/3485114.3485122. `Actions` and `SceneStates` are related tightly because an Action is conceptually the object that would change the current state to a new state. 

There is generally a distinction between robotic actions vs non-robotic actions. The difference is that the robotic ones require motion planning, which is fulfilled by the motion planning backend of compas_fab. The distinction is implemented by inheriting from the `RoboticAction` class. A number of common Actions are provided in the compas_fab library, but users are free to implement their own actions according to the needs of their process and the function of their tools.

In any robotic process, the initial state is described with `SceneState` object. It contains the state of all the mutable properties of objects in the planning scene. The objects include `RobotState`, `ToolState` (multiple), and WorkpieceState (multiple). The default implementation of these objects contains a minimal set of properties. For example, WorkpieceState contains the location (`Frame`) of the workpiece. Users can inherit these as base classes and add more properties if needed. In addition, the `SceneState` contains information describing tool attachment (to Robot) and workpiece attachment (to Tool). See API of SceneState for more details.

As actions are performed (or applied in simulation), they influence the state of the scene, resulting in a new scene state. The new scene state can be used to describe the beginning of the next action and so on. This function to modify a starting state to create an ending state is encoded using the function `Action.apply_to(self, scene_state,)`. This function allows automatic parsing of all intermediate states once the (1) initial state and (2) action list are created. This allows the user to visualize the scene, where object locations and attachment relationships can be examined. This visualization can happen even before motion planning and can help identify incorrect modeling that may cause motion planning failure.

During motion planning, the planner is preloaded with all the object models. The input to plan a single RoboticAction is the starting scene state and the RoboticAction object. In special cases where non-sequential planning is required, the SceneState at the beginning of any action can be computed automatically, allowing a single Action to be planned out of sequence. See Tutorial: Non-Sequential Planning (not yet written) for more details.

After motion planning, the planned trajectory will be stored in the same Action object. Which can be used for visualization and execution. 

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas_fab.robots.CollisionMesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if appropriate)

yck011522 committed:
Initial intro of Action and State classes

yck011522 committed:
doc string wip

yck011522 committed:
Merge remote-tracking branch 'compas-dev/main' into introduce_actions

yck011522 committed:
Refactor SceneState, WorkpieceState, ToolState and
RobotState classes

yck011522 committed:
Improved documentation of Planning Module

yck011522 committed:
Fix formatting and typos in planning module

yck011522 committed:
Added more details to Linear and Free Movement

yck011522 committed:
CartesianMovement

yck011522 committed:
Naming change

yck011522 committed:
Renaming

yck011522 committed:
lint

yck011522 committed:
tests

yck011522 committed:
lint

yck011522 commented:
Link to https://github.com/compas-dev/compas/pull/1209 for potential synergy.

yck011522 committed:
HideWorkpieces and ShowWorkpieces Actions

yck011522 committed:
Change Naming: CartesianMotion to LinearMotion

yck011522 committed:
Update LinearMotion class to include multiple
linear segments by specifying intermediate_targets.

yck011522 committed:
missing HideWorkpieces and ShowWorkpieces in action.all

yck011522 committed:
Add doc to Action Class about check preconditions and apply effects

yck011522 committed:
update test

yck011522 commented:
@gonzalocasas @chenkasirer Will you have time to take a look at this?

gonzalocasas commented on src/compas_fab/planning/action.py:
```diff
@@ -0,0 +1,765 @@
+    ----------
+    intermediate_planning_waypoint : list(:class:`compas.robots.Configuration`)
+        List of configurations to define a multi-step free movement.
+        The first and last configuration (the starting and target configuration) is not required.
+        Only include the intermediate waypoints.
```
The need for this comment is exactly what makes it odd, it's not really intuitive that you need to strip parts of the data for this input param.

gonzalocasas reviewed (commented):
It's looking good! I added a bunch of feedback, the only thing that really worries me is the lack of support for multiple robots, which I feel might be an impossible thing to add in the future, so, as mentioned, let's chat about it soon and see what are the options.

yck011522 commented on src/compas_fab/planning/__init__.py:
```diff
@@ -0,0 +1,105 @@
+"""
```
Yea, I think I can do that whole restyling thing perhaps in a month or two after major changes are settled. I'll just open a PR and remind myself.

yck011522 commented on src/compas_fab/planning/__init__.py:
```diff
@@ -0,0 +1,105 @@
+    FreeMotion
+    OpenGripper
+    CloseGripper
+    ManuallyMoveWorkpiece
```
Good idea.

yck011522 committed:
Apply suggestions (docstrings) from code review

Co-authored-by: Gonzalo Casas <casas@arch.ethz.ch>

yck011522 committed:
Rename ManuallyMoveWorkpiece to ManualWorkpieceMotion

yck011522 committed:
Fixed Missing autosummary classes

yck011522 commented on src/compas_fab/planning/state.py:
```diff
@@ -0,0 +1,280 @@
+        workpiece_id="undefined_workpiece",
+        frame=None,
+        attached_to_robot=False,
+        attached_to_robot_grasp=Transformation(),
```
Okay, I'll use None as default, that is more consistent with the attached_to_robot=False

yck011522 committed:
fix multiple import per line

yck011522 committed:
Better Doc String for States,
Test adding an example to WorkpieceState

yck011522 committed:
Fix import WorkpieceState error in example oode

yck011522 commented on src/compas_fab/planning/state.py:
```diff
@@ -0,0 +1,280 @@
+class SceneState(Data):
+    """Container for the states of all workpieces, tools and robot in a static scene.
+
+    Current implementation supports multiple workpieces and tools, but only one robot.
```
For the record as we have discussed in our meeting: the scene state will only be able to keep track of one single robot (aka one single robot model / URDF model) . 

For Complex robots like the RFL, the current workflow with a compas_fab model can support multiple planning groups. All the backends can also support multiple planning groups. It will be the role of Actions and planning calls to keep track of which planning group to use.

yck011522 committed:
Try something else with the example code

yck011522 committed:
Update src/compas_fab/planning/action.py

Co-authored-by: Gonzalo Casas <casas@arch.ethz.ch>

yck011522 committed:
Added Planning Group as attribute. Remove planned results.
Moved intermediate_planning_waypoint to FreeMotion

RobotAction.apply_effects does not take the robot config from ActionPlanResult yet, That will come in next PR

yck011522 committed:
docstring typo

yck011522 commented:
Lets hold on to resolve the Targets #409 before going back to this one.

gonzalocasas commented:
@yck011522 also, this is already done in `main` now, right?
